code/game/objects/items/devices/scanners/sequence_scanner.dm

```diff
@@ -15,65 +15,78 @@
 	add_fingerprint(user)
-	if (!HAS_TRAIT(M, TRAIT_GENELESS) && !HAS_TRAIT(M, TRAIT_BADDNA)) //no scanning if its a husk or DNA-less Species
-		user.visible_message(span_notice("[user] analyzes [M]'s genetic sequence."), \
-							span_notice("You analyze [M]'s genetic sequence."))
-		gene_scan(M, user)
-
+	//no scanning if its a husk or DNA-less Species
+	if (!HAS_TRAIT(mob, TRAIT_GENELESS) && !HAS_TRAIT(mob, TRAIT_BADDNA))
+		user.visible_message(span_notice("[user] analyzes [mob]'s genetic sequence."), \
+							span_notice("You analyze [mob]'s genetic sequence."))
+		gene_scan(mob, user)
 	else
-		user.visible_message(span_notice("[user] fails to analyze [M]'s genetic sequence."), span_warning("[M] has no readable genetic sequence!"))
+		user.visible_message(span_notice("[user] fails to analyze [mob]'s genetic sequence."), span_warning("[mob] has no readable genetic sequence!"))
 
 /obj/item/sequence_scanner/attack_self(mob/user)
 	display_sequence(user)
 
 /obj/item/sequence_scanner/attack_self_tk(mob/user)
 	return
 
-/obj/item/sequence_scanner/afterattack(obj/O, mob/user, proximity)
+/obj/item/sequence_scanner/afterattack(obj/object, mob/user, proximity)
 	. = ..()
-	if(!istype(O) || !proximity)
+	if(!istype(object) || !proximity)
 		return
 
-	if(istype(O, /obj/machinery/computer/scan_consolenew))
-		var/obj/machinery/computer/scan_consolenew/C = O
-		if(C.stored_research)
+	if(istype(object, /obj/machinery/computer/scan_consolenew))
+		var/obj/machinery/computer/scan_consolenew/console = object
+		if(console.stored_research)
 			to_chat(user, span_notice("[name] linked to central research database."))
-			discovered = C.stored_research.discovered_mutations
+			discovered = console.stored_research.discovered_mutations
 		else
 			to_chat(user,span_warning("No database to update from."))
 
-/obj/item/sequence_scanner/proc/gene_scan(mob/living/carbon/C, mob/living/user)
-	if(!iscarbon(C) || !C.has_dna())
+/obj/item/sequence_scanner/proc/gene_scan(mob/living/carbon/mob, mob/living/user)
+	if(!iscarbon(mob) || !mob.has_dna())
 		return
-	buffer = C.dna.mutation_index
-	to_chat(user, span_notice("Subject [C.name]'s DNA sequence has been saved to buffer."))
-	if(LAZYLEN(buffer))
-		for(var/A in buffer)
-			to_chat(user, span_notice("[get_display_name(A)]"))
 
+	//add mob mutations to list as well as extra mutations.
+	//dupe list as scanner could modify mob data
+	buffer = LAZYLISTDUPLICATE(mob.dna.mutation_index)
+	var/list/active_mutations = list()
+	for(var/datum/mutation/human/mutation in mob.dna.mutations)
+		LAZYOR(buffer, mutation.type)
+		active_mutations.Add(mutation.type)
+
+	to_chat(user, span_notice("Subject [mob.name]'s DNA sequence has been saved to buffer."))
+	if(LAZYLEN(buffer))
+		for(var/mutation in buffer)
+			//highlight activated mutations
+			//if(mob.dna.is_gene_active(mutation)) //doesnt work on extra mutations, might have to modify the proc itself
```

Comment:
Your current solutions seems fine,. is_gene_active doesn't work because extra mutations can't be inactive and thus isn't check for

Reply:
I see a lot of doubled and redundant stuff in there. In my opinion DNA Scanner should ask mob for active mutations and not do it itself. But I prefer to push changes continually rather then making a big PR. It will allow code to be upgraded in real-time and there is less risk of getting PR rejected if I get the wrong idea.

Reply:
In my defense I wrote genetics when I was 14-15, it doesn’t make sense

Reply:
Dont worry about it. Doing stuff like this in this age is a good feat. I saw code made by "professionals" that was 10x worse than that and it was working. Until you had to make even a tiniest change.